lib/Wikia/src/Tracer/WikiaTracer.php

```diff
@@ -0,0 +1,214 @@
+		$serverHeaderName = strtoupper( $serverHeaderName );
+		$serverHeaderName = 'HTTP_' . $serverHeaderName;
+
+		return isset( $_SERVER[$serverHeaderName] ) && $_SERVER[$serverHeaderName] !== ''
+			? $_SERVER[$serverHeaderName] : null;
+	}
+
+	private function updateContext() {
+		$newContext = array_merge(
+			$this->getApplicationContext(),
+			$this->removeNullEntries( [
+				'client_ip' => $this->clientIp,
+				'client_beacon_id' => $this->clientBeaconId,
+				'client_device_id' => $this->clientDeviceId,
+				'user_id' => $this->userId,
+				'trace_id' => $this->traceId,
+			] )
+		);
+		if ( $this->contextSource->getContext() !== $newContext ) {
+			$this->contextSource->setContext( $newContext ); // this will notify listeners
+		}
+	}
+
+	private function getApplicationContext() {
+		global $wgDBname, $wgCityId, $maintClass;
+
+		$context = [ ];
+
+		if ( !empty( $wgDBname ) ) {
+			$context['wiki_dbname'] = $wgDBname;
+		}
+
+		if ( !empty( $wgCityId ) ) {
+			$context['wiki_id'] = $wgCityId;
+		}
+
+		if ( isset( $_SERVER['REQUEST_URI'] ) ) {
+			$context['http_url_path'] = $_SERVER['REQUEST_URI'];
+
+			if ( isset( $_SERVER['REQUEST_METHOD'] ) ) {
+				$context['http_method'] = $_SERVER['REQUEST_METHOD'];
+			}
+
+			if ( isset( $_SERVER['SERVER_NAME'] ) ) {
+				$context['http_url_domain'] = $_SERVER['SERVER_NAME'];
+
+				$context['http_url'] = sprintf( "%s://%s%s",
+					( empty( $_SERVER['HTTPS'] ) ? 'http' : 'https' ),
+					$_SERVER['SERVER_NAME'],
+					$_SERVER['REQUEST_URI'] );
+			}
+
+			if ( isset( $_SERVER['HTTP_REFERER'] ) ) {
+				$context['http_referrer'] = $_SERVER['HTTP_REFERER'];
+			}
+		}
+
+		// add some context for maintenance scripts
+		if ( defined( 'RUN_MAINTENANCE_IF_MAIN' ) ) {
+			if ( isset( $_SERVER['SCRIPT_FILENAME'] ) ) {
+				$context['maintenance_file'] = realpath( $_SERVER['SCRIPT_FILENAME'] );
+			}
+
+			if ( !empty( $maintClass ) ) {
+				$context['maintenance_class'] = $maintClass;
+			}
+		}
+
+		return $this->removeNullEntries( $context );
+	}
+
+	private function removeNullEntries( $array ) {
+		foreach ( $array as $k => $v ) {
+			if ( is_null( $v ) ) {
+				unset( $array[$k] );
+			}
+		}
+
+		return $array;
+	}
+
+	/**
+	 * Update internal state from Mediawiki
+	 *
+	 * @return true so it can be used as hook handler
+	 * @throws \MWException
+	 */
+	public static function updateInstanceFromMediawiki() {
+		self::instance()->updateFromMediawiki();
+
+		return true;
+	}
+
+	public function updateFromMediawiki() {
+		global $wgRequest, $wgUser;
+		if ( is_null( $this->clientIp ) && $wgRequest && $wgRequest->getIP() ) {
+			$this->clientIp = $wgRequest->getIP();
+		}
+		if ( is_null( $this->clientBeaconId ) && $this->getRequestHeader( self::LEGACY_BEACON_HEADER_NAME ) ) {
+			$this->clientBeaconId = $this->getRequestHeader( self::LEGACY_BEACON_HEADER_NAME );
+		}
+		if ( is_null( $this->userId ) && $wgUser && $wgUser->isLoggedIn() ) {
+			$this->userId = (string)$wgUser->getId();
```

Comment:
According to our API guidelines all IDs should be strings, just in case we decide we need to put some additional info in something that used to be plain number.